File: models/Gpt-3.5-turbo/py/code/additional_problems/Permutation_Inversions.py

MOD = int(1e9) + 7

def count_permutations(n, k):
    dp = [[0] * (k+1) for _ in range(n+1)]
    dp[1][0] = 1

    for i in range(2, n+1):
        dp[i][0] = 1
        for j in range(1, k+1):
            dp[i][j] = (dp[i][j-1] + dp[i-1][j]) % MOD
            if j >= i:
                dp[i][j] = (dp[i][j] - dp[i-1][j-i] + MOD) % MOD

    return dp[n][k]

n, k = map(int, input().split())
print(count_permutations(n, k))

File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Permutation_Inversions.cpp

#include <iostream>
#include <vector>

const int MOD = 1000000007;

int countPermutations(int n, int k) {
    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(k + 1, 0));

    dp[1][0] = 1;

    for (int i = 2; i <= n; ++i) {
        dp[i][0] = 1;
        for (int j = 1; j <= k; ++j) {
            dp[i][j] = (dp[i][j] + dp[i - 1][j]) % MOD;
            for (int l = 1; l <= i - 1 && l <= j; ++l) {
                dp[i][j] = (dp[i][j] + dp[i - 1][j - l]) % MOD;
            }
        }
    }

    return dp[n][k];
}

int main() {
    int n, k;
    std::cin >> n >> k;

    std::cout << countPermutations(n, k) << std::endl;

    return 0;
}
